libraries/bot-applicationinsights/src/test/java/com/microsoft/bot/applicationinsights/TelemetryWaterfallTests.java

```diff
@@ -0,0 +1,396 @@
+
+            }
+
+            @Override
+            public void trackDependency(String dependencyTypeName, String target, String dependencyName, String data, OffsetDateTime startTime, Duration duration, String resultCode, boolean success) {
+
+            }
+
+            @Override
+            public void trackEvent(String eventName, Map<String, String> properties, Map<String, Double> metrics) {
+                StringBuilder sb = new StringBuilder(eventName).append("_").append(counter[0]++);
+                saved_properties.put(sb.toString(), properties);
+            }
+
+            @Override
+            public void trackException(Exception exception, Map<String, String> properties, Map<String, Double> metrics) {
+
+            }
+
+            @Override
+            public void trackTrace(String message, Severity severityLevel, Map<String, String> properties) {
+
+            }
+
+            @Override
+            public void trackDialogView(String dialogName, Map<String, String> properties, Map<String, Double> metrics) {
+
+            }
+
+            @Override
+            public void flush() {
+
+            }
+        };
+
+        MyWaterfallDialog waterfallDialog = new MyWaterfallDialog("test", newWaterfall());
+
+        dialogs.add(waterfallDialog);
+        dialogs.setTelemetryClient(telemetryClient);
+
+        new TestFlow(adapter, turnContext -> {
+            DialogContext dc = dialogs.createContext(turnContext).join();
+            dc.continueDialog().join();
+            if (!turnContext.getResponded()) {
+                dc.beginDialog("test", null).join();
+            }
+            return CompletableFuture.completedFuture(null);
+        })
+        .send("hello")
+        .assertReply("step1")
+        .send("hello")
+        .assertReply("step2")
+        .send("hello")
+        .assertReply("step3")
+        .send("hello")
+        .assertReply("step1")
+        .startTest()
+        .join();
+
+        Assert.assertEquals((Integer)7, counter[0]);
+
+        // Verify:
+        // Event name is "WaterfallComplete"
+        // Event occurs on the 4th event logged
+        // Event contains DialogId
+        // Event DialogId is set correctly.
+        Assert.assertTrue(saved_properties.get("WaterfallComplete_4").containsKey("DialogId"));
+        Assert.assertEquals("test", saved_properties.get("WaterfallComplete_4").get("DialogId"));
+        Assert.assertTrue(saved_properties.get("WaterfallComplete_4").containsKey("InstanceId"));
+        Assert.assertTrue(saved_properties.get("WaterfallStep_1").containsKey("InstanceId"));
+
+        // Verify naming on lambda's is "StepXofY"
+        Assert.assertTrue(saved_properties.get("WaterfallStep_1").containsKey("StepName"));
+        Assert.assertEquals("Step1of3", saved_properties.get("WaterfallStep_1").get("StepName"));
+        Assert.assertTrue(waterfallDialog.getEndDialogCalled());
```

Comment:
Are you missing this [assert](https://github.com/microsoft/botbuilder-dotnet/blob/main/tests/Microsoft.Bot.Builder.ApplicationInsights.Tests/TelemetryWaterfallTests.cs#L154)?

Reply:
No, in C# there is an extra assert that is [duplicated](https://github.com/microsoft/botbuilder-dotnet/blob/main/tests/Microsoft.Bot.Builder.ApplicationInsights.Tests/TelemetryWaterfallTests.cs#L149). Since there is no assignation or any change of state at this point of the test, I didn't add it because the assert was duplicated.

Reply:
Great